### ISSUE TITLE
[rv_dm] V2S Signoff

### ISSUE THREAD

**Issue by andreaskurth** [lowRISC/opentitan#22472]

**Comment by moidx** [comment]
Discussed moving as P1 to M5 due to RTL risk analysis performed during last issue triage.

**Comment by rswarbrick** [comment]
I'm lowering the effort estimate for this issue, so that it tracks "the effort to do the signoff, once we've implemented stuff" (where the implementation is tracked in separate issues).

**Comment by rswarbrick** [comment]
Here are some notes about V2 status

The current state of rv_dm is not quite ready for the V2 development
stage. Dividing into checklist items gives the following:

- `DESIGN_DELTAS_CAPTURED_V2`
  + The DV document is pretty bare bones so capturing any changes
    there is really covered by a different checklist item.
  + On the testplan side, this is mostly in place except for the fact
    that the new NDM/WARL behaviour is not currently planned. This is
    tracked in issue #22048.
  + Estimated effort to completion: **1 day** for testplan changes only
- `DV_DOC_COMPLETED`
  + There is *some* documentation in `hw/ip/rv_dm/dv/README.md` but
    it's definitely not complete. Indeed, there are actually still
    placeholder lines left in the "assertions" section.
  + Estimated effort to completion: **1 day**
- `FUNTIONAL_COVERAGE_IMPLEMENTED`
  + There is no functional coverage plan, so this is true in a vacuous
    sense. This is probably not the right place to track the fact it
    is missing.
- `ALL_INTERFACES_EXERCISED`
  + This is true.
- `ALL_ASSERTION_CHECKS_ADDED`
  + Again, there are no "planned assertions": so this checklist item
    is vacuously true.
  + On the plus side, we *do* have assertions about all bus interfaces
    that one might expect.
- `SIM_TB_ENV_COMPLETED`
  + This is true.
- `SIM_ALL_TESTS_PASSING`
  + This is not true for a few reasons.
    1. There is a feature that is not currently covered in the
       testplan (nor is it tested). Tracked in #22044.
    2. There is a V2 testpoint that isn't currently implemented in the
       repository. Fortunately, there is a PR that implements it
       (#23992), which should resolve this.
    3. There are a series of six unimplemented V2S testpoints
       (`sec_cm_lc_dft_en_intersig_mubi`,
       `sec_cm_otp_dis_rv_dm_late_debug_intersig_mubi`,
       `sec_cm_dm_en_ctrl_lc_gated`,
       `sec_cm_sba_tl_lc_gate_fsm_sparse`,
       `sec_cm_mem_tl_lc_gate_fsm_sparse`,
       `sec_cm_exec_ctrl_mubi`).
    4. All the tests that are implemented are passing for most seeds
       except `rv_dm_stress_all_with_rand_reset` which fails for every
       seed.
  + I believe that we normally interpret this checklist item as being
    confined to testpoints that are V1 or V2. If so pass rates for
    existing tests mean that, this will actually be completed for them
    once #23992 is merged.
  + The first item on the list above is not implemented. Estimated effort there is **4 days**.
- `SIM_NIGHTLY_REGRESSION_V2`
  + This is satisfied. Even including the broken
    `rv_dm_stress_all_with_rand_reset` test, the pass rate last night
    was 97.18%.
- `SIM_CODE_COVERAGE_V2`:
  + This is not satisfied.
  + Estimated effort to completion: **5 days**
- `SIM_FUNCTIONAL_COVERAGE_V2`:
  + Despite the fact that there is no functional coverage defined for
    this block, the checklist item *still* isn't satisfied!
  + The missing coverage (defined but not covered) turns out to be
    from two things:
    - `mubi_cg` for `late_debug_enable` (which is a 32-bit mubi
      register with a rather enormous array of "others" coverpoint
      bins).
    - `max_outstanding_cg` for the TL interface. I strongly suspect
      that the problem here is that we didn't configure the TL agent
      config appropriately.
  + Taking this literally, estimated effort to completion: **1 day**.
- `PRE_VERIFIED_SUB_MODULES_V2`
  + True in a vacuous sense: there are no pre-verified submodules.
- `SEC_CM_PLANNED`
  + This is true.
- `NO_HIGH_PRIORITY_ISSUES_PENDING`, `ALL_LOW_PRIORITY_ISSUES_ROOT_CAUSED`:
  + This is true. There are currently no open design-related issues.
- `DV_DOC_TESTPLAN_REVIEWED`
  + The DV document hasn't been reviewed (or probably completed: see above!).
  + The testplan has been reviewed, and rewritten in places as it was
    implemented.

I'm skipping `FPV_ALL_ASSERTIONS_WRITTEN`,
`FPV_ALL_ASSUMPTIONS_REVIEWED`, `SIM_FW_SIMULATED`,
`FPV_CODE_COVERAGE_V2`, `FPV_COI_COVERAGE_V2`.


Adding the items above, my effort estimate for getting ready for V2
status is **12 days** work. Note that V2S is further effort on top of
this (but probably not a vast amount).

**Comment by johngt** [comment]
Based on the approval of the TC this effort is deferred to at least M6.
Making the Milestone change now.
@moidx / @vogelpi  / @andreaskurth

**Comment by rswarbrick** [comment]
Some of the work for V2 (rather than V2S) has now been done. Remaining is:
- `DV_DOC_COMPLETED`. Estimate: 1 day.
- `SIM_CODE_COVERAGE_V2`. Some of this is UNR (see #24146). After that is fixed, it doesn't look like we are far from the threshold. Estimate: 1 day.
- `SIM_FUNCTIONAL_COVERAGE_V2`. The TL coverage problem has a PR (#24141). Remaining is the mubi coverage, which seems to be a bug in our base coverage code (tracked as #24142). estimate: 1 day.
- `DV_DOC_TESTPLAN_REVIEWED`. Document probably needs a review once it is complete. Estimate: 0.5 days.

Summing the estimates gives 3.5 days remaining (to V2)

**Comment by rswarbrick** [comment]
I've just realised that the estimate for this issue is for V2 signoff (rather than V2S, which will be slightly more). On the V2 side, I think there is still 3.5 days effort remaining for V2.

Estimating the remaining work required once V2 is signed off:

- `SEC_CM_TESTPLAN_COMPLETED`: This is done.
- `SIM_SEC_CM_VERIFIED`: This is essentially the work tracked by #23721. Fortunately, there's not much work left for that. There are now only three testpoints unimplemented (two sparse FSM testpoints and one mubi testpoint). Estimate: 2 days.
- `SIM_COVERAGE_REVIEWED`: N/A (no waivers used for V2)
- `SEC_CM_DV_REVIEWED`: I don't think this has been performed. This is simple enough for a security expert to do on their own (e.g. @vogelpi). Estimate: 0.5 days.

Summing these times gives 2.5 days.

Adding the two development stages gives 3.5+2.5 = 6 days. Updating the tracker accordingly.

**Comment by andreaskurth** [comment]
Moving to M7 as P1, as approved by TC today

**Comment by rswarbrick** [comment]
I think things are now pretty much there for both V2 and V2S. On July 26th, I [updated](https://github.com/lowRISC/opentitan/issues/22472#issuecomment-2253589284) the estimates for remaining work for V2.

Addressing each item in turn:
- `DV_DOC_COMPLETED`: This should be done, now that #24335 is merged.
- `SIM_CODE_COVERAGE_V2`: Very annoyingly, this isn't technically done because we're a tiny bit below one of the 90% thresholds.
  + Last night's run was commit [584c3d46af](https://github.com/lowrisc/opentitan/tree/584c3d46afbad43dea506fd8055243b7e05adbce). It saw:
    ![image](https://github.com/user-attachments/assets/05a104eb-00db-4909-8983-e98e0afa231d)
   + The numbers get a little bit better still with some commits merged today, but I'm locally seeing a COND coverage of 89.39% (aargh!)
   + The remaining gaps should all be small nits. I've gone through the coverage report in detail and I think that none of the gaps is of interest.
- `SIM_FUNCTIONAL_COVERAGE_V2`: The rv_dm testbench should make this trivial to hit because it doesn't define many covergroups. Unfortunately, there's what appears to be a VCS bug in merging the groups of `mubi32_cov::mubi_cg`. I've tried a few things locally (and merged one of them, which had no effect), but it generates vast numbers of coverage bins that cannot be hit, as part of merging runs. The result looks like this:
  ![image](https://github.com/user-attachments/assets/21a85b63-666d-4c44-afdb-8162f170588d)
  + Looking manually at the report, one version of each bin in the first covergroup is hit:
    ![image](https://github.com/user-attachments/assets/98d47709-bcc7-42a7-ba25-3da772defc6d)
  + Considering the other bins as caused by a tooling bug, the only bin that is not being it is `false` (caused by the fact that my DV code selects a random value when it is trying to choose something that isn't MuBi32True). Even with this hole hole, the numbers are still well over the threshold.
  + Last night's report shows 348 bins in total for this coverpoint, when there should have been 32+2 = 34:
     ![image](https://github.com/user-attachments/assets/b0790e04-ba99-4ca3-bfa9-13e741d62cd2)
   + The groups coverage summary is currently this:
    ![image](https://github.com/user-attachments/assets/dec1191b-e976-4d36-80a0-b4cd578255e7)
   + Subtracting 348-34 = 314 from the "expected" number in the summary gives 446 groups covered of 763-318=449 expected, for a ratio of 99.3%
- `DV_DOC_TESTPLAN_REVIEWED`: The testplan has been reviewed and I've talked through the DV document with @hcallahan-lowrisc, who also reviewed the deltas in the PR that was merged this afternoon.

I calculated the remaining V2S work items in a [comment](https://github.com/lowRISC/opentitan/issues/22472#issuecomment-2256385495) on July 29th. Items that remained were:
- `SIM_SEC_CM_VERIFIED`: The testplan is implemented and passes for all seeds.
- `SEC_CM_DV_REVIEWED`: This has not been performed yet, but we could ask @vogelpi to take a look over it as part of signing off this development stage.

Thus, I believe that the remaining work is:
- Either waive the delta or further improve code coverage numbers. (From a day's experience, the results are *very* incremental)
- Explicitly waive the functional coverage hole, which is pretty obviously caused by a tooling bug.
- Get a review of the security DV (possibly by @vogelpi as part of this signoff).

**Comment by andreaskurth** [comment]
Just discussed the coverage numbers that are nominally below the V2 threshould (COND code | functional) with @rswarbrick, and we suggest to waive the remaining percentage points if this is substantiated by the outstanding coverage gap analysis.

**Comment by vogelpi** [comment]
I've just reviewed the sec_cm testplan, here are my findings:
- All countermeasures are covered by at least one test or testpoint in the sec_cm testplan.
- Based on the detailed the testplan descriptions (thanks @rswarbrick ! this is really well done), all important aspects of the countermeasures are tested.

------

Aside from this, I find the descriptions / docs around the late debug enablement are slighly confusing (I am not particularly referring to the sec_cm testplan here!). I am not saying we need to change the documentation I write this down mostly for having a reference for myself later.

What I found most useful for understanding the implementation is the register documentation as well as the actual implementation:

>            In terms of gating conditions, the debug module implements a
>            mechanism that allows for a 'late' debug enable in the DEV life
>            cycle state:
>
>              - Debug is always enabled unconditionally in DFT-enabled life
>                cycle states (TEST_UNLOCKED* and RMA).
>              - Debug is always enabled unconditionally in the DEV life cycle
>                state iff OTP_DIS_RV_DM_LATE_DEBUG is set to kMuBi8True.
>              - Debug is conditionally enabled based on the value of the
>                LATE_DEBUG_ENABLE CSR in the DEV life cycle state iff OTP_DIS_RV_DM_LATE_DEBUG
>                is set to kMuBi8False (or any invalid MuBi value).
>              - Debug is always disabled in all other life cycle states.

```sv
  for (genvar k = 0; k < lc_ctrl_pkg::TxWidth; k++) begin : gen_mubi_mux
    assign lc_hw_debug_en_gated_raw[k] = (mubi8_test_true_strict(otp_dis_rv_dm_late_debug[k]) ||
                                          mubi32_test_true_strict(late_debug_enable[k])) ?
                                          lc_hw_debug_en_raw[k] :
                                          lc_dft_en_raw[k];
  end
```

- So, if `otp_dis_rv_dm_late_debug` is strictly true, we use `lc_hw_debug_en` (only true in DEV).
- If `otp_dis_rv_dm_late_debug`  is false (late debug enable possible), we use  `lc_dft_en` (which is only true in TEST_UNLOCKED* and RMA but not in DEV) until software writes the `late_debug_enable` CSR when we switch to `lc_hw_debug_en` (again only true in DEV).

**Comment by rswarbrick** [comment]
Following up on this, I'm pleased to say that (minor!) fiddling around has now got the coverage numbers above the thresholds for V2. For example, the most recent run was 24/09/24 with the following coverage:
![image](https://github.com/user-attachments/assets/6579c16e-5550-4b6c-b00a-ccb148f2cb98)

As far as I can see, we have now cleared the (admittedly low) bar for V2. Based on the message from @vogelpi, I think we can also assert V2S. Pirmin, does that sound right to you?

**Comment by vogelpi** [comment]
Yes, that sound right. Nice work @rswarbrick !

In case someone else is wondering about the group coverage, it's a tool bug: VCS fails to merge the coverage of the `mubi32_cov::mubi_cg` groups. Rupert explained this above.

**Comment by rswarbrick** [comment]
Thanks. I'll file a PR in a minute that bumps up the claimed version.
